woocommerce/woocommerce-gateway-stripe
# Add multisite support and notify merchants after acccount changes

Fixes #1092.

## Changes proposed in this Pull Request:

Many merchants are experiencing errors, related to missing customers. Those errors can be caused by multiple factors:

1. The Stripe account was changed.
2. The customer was manually deleted in Stripe.
3. The extension is used in a multisite network.

While No. 2 is a conscious decision, 1 & 3 are addressed by this PR by the following changes:

### Switching to `*_user_option` instead of `*_user_meta`

H/T @thenbrent for suggesting those APIs!

The `*_user_option` APIs are built on top of user meta APIs, but add blog-specific prefixes to meta keys in order to isolate bits of data for a specific site.

`get_user_option()` also has a built-in fallback in case a setting is specified globally, but not for the given site. This means that the new version of the gateway will not degrade UX any further, but might improve it with time. There is also [a section in this doc](https://docs.woocommerce.com/document/stripe-fixing-customer-errors/#section-5), which describes how to transfer the existing values to the main site.

### Adding a notice upon changes in API keys

The new `WC_Gateway_Stripe::process_admin_options` method (which overloads the one from `WC_Settings_API`) compares the keys before & after saving them, and adds a notice to the admin. The notice will not be displayed when those values are added for the first time (when transitioning from an empty string).

The content of the notice leads merchants to https://docs.woocommerce.com/document/stripe-fixing-customer-errors/#section-1, where I've described how to clean up the database if necessary.

### Improving the content of "no such customer" messages

[Here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1131/files#diff-f8dd0b04c52dc566825dead0a3b8b84cR187) I've added additional strings to the message, which suggest that there might be changes in the Stripe environment.

Please note that those are relatively redundant, since Stripe already changes their message if a test/live user exists with the same ID. Still, this should customers to consider whether they have changed something.

## Testing instructions

There are no particular instructions, you need to make purchases with new & existing customers and everything should keep on working well.

While doing so, please make sure to:

- Try it with different accounts. This will trigger the new notice, as well as some "No such customer" errors.
- Try it within a multisite environment, as well as a single site.

-------------------
